src/libexpr/primops.cc

```diff
@@ -3424,7 +3424,7 @@ static void prim_substring(EvalState & state, const PosIdx pos, Value * * args,
             .errPos = state.positions[pos]
         }));
 
-    v.mkString((unsigned int) start >= s->size() ? "" : s->substr(start, len), context);
+    v.mkString((unsigned NixInt) start >= s->size() ? "" : s->substr(start, len), context);
```

Comment:
This causes a compilation failure on macOS. Wouldn't it make more sense to cast it to `size_t`, since we're comparing to `s->size()`?

Reply:
That's annoying, a cast to size_t would overflow on 32-bit systems. I've now added a forceIntChecked function that uses `boost::numeric_cast` to do a checked cast to size_t where applicable in the primops.